.github/workflows/python-test-coverage.yml

```diff
@@ -0,0 +1,41 @@
+name: Python Test and Coverage
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Set up Python
+      uses: actions/setup-python@v3
+      with:
+        python-version: '3.11'
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install pytest pytest-cov
+        pip install -r requirements.txt
+
+    - name: Run tests with coverage
+      run: |
+        pytest --cov=src tests/
+
+    - name: Generate coverage report
+      run: |
+        pytest --cov=src --cov-report=xml tests/
+
+    - name: Upload coverage to Codecov
+      uses: codecov/codecov-action@v3
+      with:
+        token: ${{ secrets.CODECOV_TOKEN }}
```

Comment:
We would need a secrets file and a codecov account. Unfortunately, we don't have the money to do this. I posted a suggestion on how to do this for free

Reply:
Now it is failing for the reason of coverage being too low. I think it is an acceptable reason to fail a PR and can wait until we write one test per function. In fact most things should probably wait until we write one test per function.